### PR TITLE
Loki Query Editor: Autocompletion and suggestions improvements (unwrap, parser, extracted labels)

### LINKED ISSUE #57683: Loki: Extend autocompletion for unwrap and logs query in monaco editor
We want to further improve label suggestions while writing queries. For that, we should:

1. [x] Suggest labels after the unwrap operation.
2. [x] Suggest labels for logs query
3. [x] Figure out autocompletion after user selects a label, because when the user selects a label, we would need to show suggestions for operators and value. Moreover, it should be only suggested after the parser is used.

_Originally posted by @ivanahuckova in https://github.com/grafana/grafana/pull/57368#pullrequestreview-1150792543_

3. [x]  Instead of creating query from otherLabels, we should create it from the text, so it contains parser and therefore extracted keys
- Unwrap currently shouldn't offer `unwrap + label` 
- Handle cases for binary operation metric queries

_Originally posted by @ivanahuckova in https://github.com/grafana/grafana/pull/57368#pullrequestreview-1170682607_

### PR DESCRIPTION
**Pull request by matyax**
**What is this feature?**

This PR adds several improvements to our autocompletion support for the Monaco query editor.

- Improved `unwrap` identification and suggestion.
- Improved suggestions for `unwrap`: labels or functions.
- Testing and coverage improvements.
- We now send the full log query to get a data sample and extract labels, instead of just the query labels.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/57683

**Special notes for your reviewer**:

Compare previous with current suggestion behavior.

https://user-images.githubusercontent.com/1069378/203310123-d227255e-f1ef-4b82-a871-60bae3248e8f.mov

Demo of sending full logs query for extracted labels:

https://user-images.githubusercontent.com/1069378/203346970-d7106c93-070e-436f-95b9-b1f1bc291dc8.mov

### CODE REVIEW

**Review by ivanahuckova — COMMENTED** [review]
Really nice job!! 🎉 I tested this and I have couple of questions/thoughts:

> Only suggest parser when it's not present in the log query.

I am wondering if this make sense.  In most cases, you use 1 parser, but there could be cases where you want to use multiple parsers. (I think there could be, but would be nice to check with Loki team). Based on their answer, we can decide  if we want to do show parsers or not. 

> Improved unwrap identification and suggestion.

I am thinking we could move `label_format` and `line_format` above `unwrap`, because now if we have a lot of labels, these options are quite low in the list and hard to find. 

I am also thinking about the fact, that `unwrap` can be used only in metrics query. Therefore I am thinking if we should suggest it if we are in log query, as the resulting query is going to be invalid. Or if we should suggest it when we are in metrics query.

And last thing (I did a video explaining the issue), but when I write pipe `|` I get only `unwrap` suggestion and no labels/functions. 

https://user-images.githubusercontent.com/30407135/204562670-a1af035f-8426-4a76-8821-05aed395d9ef.mov

>Improved suggestions for unwrap: labels or functions.

Here I am wondering, if we should only suggest labels that are number/duration/bytes. Because for all other labels, the resulting query is invalid. And if you use parser, you can have tens of labels, but only fraction would be usable. 

And 2 more non-related-to-this PR issue I noticed - when I write queries:

- `{source="data"} | ^` and I choose parser,  it adds extra space between pipe and parser, so the resulting query is `{source="data"} |  logfmt` instead of  `{source="data"} | logfmt`. 

- `rate({source="data"} |^ [$__interval])` and I choose parser, it adds extra pipe, so the resulting query is `rate({source="data"} | | logfmt[$__interval])` instead of `rate({source="data"} |  logfmt[$__interval])`

**Comment by matyax** [comment]
Thanks for the feedback!

> here could be cases where you want to use multiple parsers.

Asked the Loki team.

> move label_format and line_format above unwrap

Good idea, will do that.

> Therefore I am thinking if we should suggest it if we are in log query, as the resulting query is going to be invalid.

That's also a good idea, but I think we can maybe let the query validation handle that?

imagen

>  I write pipe | I get only unwrap suggestion and no labels/functions.

I saw that. The problem is that when you write the pipe the situation changes to `AFTER_SELECTOR`. I'll see if we can improve that, but currently the situation identification struggles with some queries.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts [hunk 1]
```diff
@@ -256,7 +310,9 @@ export async function getCompletions(
         dataProvider
       );
     case 'AFTER_SELECTOR':
-      return getAfterSelectorCompletions(situation.labels, situation.afterPipe, dataProvider);
+      return getAfterSelectorCompletions(situation.logQuery, situation.afterPipe, situation.parser, dataProvider);
+    case 'AFTER_UNWRAP':
```
[the remark this answers is not in the document]
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:314]: You get the, but the limitation is that you have to be in the process of writing the query:
`quantile_over_time(0.99, {cluster="ops-tools1",container="ingress-nginx"} | json | __error__ = "" | unwrap `
imagen

If you add the closing brace, the situation changes and it no longer identifies it.

`quantile_over_time(0.99, {cluster="ops-tools1",container="ingress-nginx"} | json | __error__ = "" | unwrap )`

imagen

I'll see if I can improve the situation detection there.

**Comment by matyax** [comment]
@ivanahuckova could you please take this for a spin again? Thank you 🙇

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts [hunk 2]
```diff
@@ -116,7 +119,12 @@ export type Situation =
   | {
       type: 'AFTER_SELECTOR';
       afterPipe: boolean;
-      labels: Label[];
+      parser?: string;
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts:122]: Do we still need this parser here? I thought that we had it here to not offer 2 parsers, but we learned that there might be many parsers. So do we still need it?

**Review by ivanahuckova — COMMENTED** [review]
The line format and label format are still at the bottom 
image


Moreover is there a plan to offer only number/duration/bytes unwrap terms in this PR of in next?

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts (hunk 2 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts]: No, thanks for noticing.

**Comment by matyax** [comment]
@ivanahuckova Let's use a follow-up PR to address:
- Move up line_format and label_format 
- Resolve the extra added space.

> Moreover is there a plan to offer only number/duration/bytes unwrap terms in this PR of in next?

It's hard to tell what query are you using there, but we do suggest them in metrics query such as: `sum(sum_over_time({place="luna"} | unwrap ^ [5m])) by (place)

imagen

> and I choose parser, it adds extra pipe, so the resulting query is 

I haven't been able to consistently reproduce this with input. So far it seems to randomly pop up.

**Review by ivanahuckova — APPROVED** [review]

**Comment by matyax** [comment]
@ivanahuckova 

> I meant, that currently we offer all labels. But unwrap can be only with label that's value is number/duration/bytes type

From https://grafana.com/docs/loki/latest/logql/metric_queries/ I read that 

> The unwrap expression is noted | unwrap label_identifier

And

> Optionally the label identifier can be wrapped by a conversion function | unwrap <function>(label_identifier)

Can you confirm if what you mean is that we're missing autocompletion for `<function>(label_identifier)`?

**Comment by ivanahuckova** [comment]
No, I meant this part 

```
The unwrap expression is noted | unwrap label_identifier where the label identifier is the label name to use for extracting sample values.

Since label values are string, by default a conversion into a float (64bits) will be attempted, in case of failure the __error__ label is added to the sample. 
``` 

So if we offer `unwrap level` and level is string that can't be converted into float number, user will get error. So we don't want to offer labels that have non-number values, as result of this query is error.

**Comment by matyax** [comment]
Oh, I remember that. I'm not really sure if we can resolve that in the autocomplete layer (specially without adding delay as the user types). I'll explore it with the follow up.

**Pull request merged**